source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc

```diff
@@ -0,0 +1,111 @@
+#include "extensions/filters/common/ext_authz/ext_authz_http_impl.h"
+
+#include "common/common/enum_to_int.h"
+#include "common/http/async_client_impl.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace ExtAuthz {
+
+namespace {
+
+const Http::HeaderMap* getZeroContentLengthHeader() {
+  static const Http::HeaderMap* header_map =
+      new Http::HeaderMapImpl{{Http::Headers::get().ContentLength, std::to_string(0)}};
+  return header_map;
+}
+
+} // namespace
+
+RawHttpClientImpl::RawHttpClientImpl(const std::string& cluster_name,
+                                     Upstream::ClusterManager& cluster_manager,
+                                     const absl::optional<std::chrono::milliseconds>& timeout,
+                                     const std::string& path_prefix)
+    : cluster_name_(cluster_name), path_prefix_(path_prefix), timeout_(timeout),
+      cm_(cluster_manager) {}
+
+RawHttpClientImpl::~RawHttpClientImpl() { ASSERT(!callbacks_); }
+
+void RawHttpClientImpl::cancel() {
+  ASSERT(callbacks_ != nullptr);
+  request_->cancel();
+  callbacks_ = nullptr;
+}
+
+void RawHttpClientImpl::check(RequestCallbacks& callbacks,
+                              const envoy::service::auth::v2alpha::CheckRequest& request,
+                              Tracing::Span&) {
+  ASSERT(callbacks_ == nullptr);
+  callbacks_ = &callbacks;
+
+  Http::HeaderMapPtr headers = std::make_unique<Http::HeaderMapImpl>(*getZeroContentLengthHeader());
+  for (const auto& header : request.attributes().request().http().headers()) {
+    const Http::LowerCaseString key{header.first};
+    if (key == Http::Headers::get().Path && !path_prefix_.empty()) {
+      std::string value;
+      absl::StrAppend(&value, path_prefix_, header.second);
+      headers->addCopy(key, value);
+    } else {
+      headers->addCopy(key, header.second);
+    }
+  }
+
+  request_ = cm_.httpAsyncClientForCluster(cluster_name_)
+                 .send(std::make_unique<Envoy::Http::RequestMessageImpl>(std::move(headers)), *this,
+                       timeout_);
+}
+
+void RawHttpClientImpl::onSuccess(Http::MessagePtr&& response) {
+  ResponsePtr authz_response = std::make_unique<Response>(Response{});
+
+  uint64_t status_code;
+  if (StringUtil::atoul(response->headers().Status()->value().c_str(), status_code)) {
+    if (status_code == enumToInt(Http::Code::OK)) {
+      response->headers().removeStatus();
+      response->headers().removeDate();
+      response->headers().removeEnvoyUpstreamServiceTime();
+      response->headers().removeMethod();
+      response->headers().removePath();
+      response->headers().removeContentLength();
+      response->headers().removeContentType();
```

Comment:
This list seems quite adhoc. How did it come into being? How we know it is complete?

Reply:
`Status, Method, Path, Content-Type and Content-Length` are most likely to be included by libraries and they might cause problems if sent to the upstream, but there is probably more. `Date and EnvoyUpstreamServiceTime` were caught during debugging, so I added them to the list.

I think that's why we had the list of allowed headers that was dropped from the design. Not sure what is best here. Suggestions?

Reply:
Isn't this the downstream? What do you mean by included by libraries? I don't see explicit content-length handling in the Envoy source, shouldn't we just forward that from the auth server response?

Reply:
These headers go to the upstream service. I meant Http servers instead of libraries.

Reply:
@htuch Thanks for looking at this. Yes, (3) is correct. I just replicated Ambassador's HTTP filter here. This behavior will be explained in the documentation for sure.